CDAP-13558 Support CDC pipelines on Spark version 2.x

mlozbin-cybervisiontech opened the pull request:
JIRA - https://issues.cask.co/browse/CDAP-13558
WIKI - https://wiki.cask.co/display/CE/CDC+Solution+Enhancements

In scope of this PR:
- implemented integration tests for Bigtable Sink plugin

mlozbin-cybervisiontech committed:
CDAP-13558 Support CDC pipelines on Spark version 2.x
- implemented integration tests for Bigtable Sink plugin

mlozbin-cybervisiontech commented:
@albertshau, could you please take a look.

albertshau commented on src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java:
```diff
@@ -0,0 +1,387 @@
+
+    @Test
+    public void testDeploymentFailedWhenCredentialsFileIsMissing() throws Exception {
+      ETLPlugin sourceConfig = MockSource.getPlugin(Schemas.CHANGE_SCHEMA, Collections.emptyList());
```
There is no need to test this through pipeline deployment. It is much less expensive to test these things in a separate CDCBigTableConfigTest class where you create instances of CDCBigTableConfig and just call validate().

mlozbin-cybervisiontech replied:
I did it for several reasons:
- to make tests look similar because we will still need config tests with real pipeline. Splitting those tests will make it more difficult to understand what exactly is covered by tests.
- to make future refactoring easier. As for me "validate" methods inside configuration DTO object violates single responsibility principle a bit.
- to test mappings (if everything is correctly mapped from properties map to config instance)

What do you think?

albertshau replied:
The only plugin specific logic that is being tested here is the validation logic in the config. Everything else is functionality that CDAP is supposed to provide, and is unit tested in the CDAP project. When possible, we just want to test plugin logic, not to test plugin logic while also testing all the stuff CDAP does.

albertshau commented on src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java:
```diff
@@ -0,0 +1,387 @@
+
+    @Before
+    public void beforeMethod() {
+      Assume.assumeNotNull(PROJECT);
```
these tests are testing validation and should always run, regardless of whether or not an actual project or instance is being provided.

albertshau commented on src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java:
```diff
@@ -0,0 +1,387 @@
+
+      Awaitility.await().atMost(TEST_TIMEOUT).pollInterval(Duration.ONE_SECOND)
+        .ignoreException(StatusRuntimeException.class).untilAsserted(() -> {
+        TableName tableName = TableName.valueOf(dbTableName);
```
nit: everything in the block should be indented

albertshau commented on src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java:
```diff
@@ -0,0 +1,387 @@
+
+      Awaitility.await().atMost(TEST_TIMEOUT).pollInterval(Duration.ONE_SECOND)
+        .ignoreException(StatusRuntimeException.class).untilAsserted(() -> {
+        TableName tableName = TableName.valueOf(dbTableName);
```
nit: everything in the block should be indented

albertshau commented on src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java:
```diff
@@ -0,0 +1,387 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@RunWith(Enclosed.class)
```
I'm not familiar with this, what is the benefit of having inner test classes instead of multiple top level test classes?

mlozbin-cybervisiontech replied:
in general it allows you to encapsulate tests with different preconditions (before/after) and some common logic/variables in one class (easier to describe clear testing conventions - one functionality - one test class).

mlozbin-cybervisiontech committed:
CDAP-13558 Support CDC pipelines on Spark version 2.x
- implemented integration tests for CTSQLServer source plugin
- introduced OperationType enum to make CDC records more type safe

mlozbin-cybervisiontech committed:
CDAP-13558 Support CDC pipelines on Spark version 2.x
- implemented integration tests for CTSQLServer source plugin
- introduced OperationType enum to make CDC records more type safe

mlozbin-cybervisiontech commented:
@albertshau, addressed all comments.

albertshau commented on src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java:
```diff
@@ -0,0 +1,22 @@
+
+  public void validate() {
+    if (!containsMacro(Constants.Reference.REFERENCE_NAME) && !References.isValid(referenceName)) {
+      String message = String.format("%s is not a valid id. Allowed characters are letters, numbers, " +
```
This message fits better in the References class, as that is where the Pattern is. I think it would be better if References.isValid() was instead References.validate() that throws an exception. This method would then catch that exception and use it's message for the InvalidConfigPropertyException.

mlozbin-cybervisiontech replied:
removed References class and avoided duplication of code from IdUtils

albertshau commented on src/main/java/co/cask/cdc/plugins/common/References.java:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
update copyright year

mlozbin-cybervisiontech replied:
removed this class and avoided duplication of code from IdUtils

albertshau commented on src/main/java/co/cask/cdc/plugins/common/CDCReferencePluginConfig.java:
```diff
@@ -0,0 +1,22 @@
+package co.cask.cdc.plugins.common;
```
missing copyright

mlozbin-cybervisiontech replied:
fixed

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -0,0 +1,134 @@
+      VALID_ACCOUNT_FILE_PATH
+    );
+
+    Assertions.assertThatCode(config::validate)
```
This can be replaced with just a call to config.validate().

mlozbin-cybervisiontech replied:
fixed

PS: I prefer BDD in cases where we want to make an accent on what is actually tested here. I understand your point and will simplify such tests in future.

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -0,0 +1,134 @@
+      VALID_ACCOUNT_FILE_PATH
+    );
+
+    Assertions.assertThatThrownBy(config::validate)
```
this should be replaced with a call to config.validate(), where you catch the exception and check that e.getProperty() is correct.

albertshau replied:
same for the next few tests

mlozbin-cybervisiontech replied:
> just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

mlozbin-cybervisiontech replied:
Added invalid property check. Can we use AssertJ here to have much more readable assertion and failure message? You can compare it with a code to make similar thing https://github.com/joel-costigliola/assertj-examples/blob/master/assertions-examples/src/test/java/org/assertj/examples/FailUsageExamples.java.
AssertJ also allows to easily extract test conditions to test framework to make tests for all plugins look similar and more readable. Currently there are no guides and multiple approaches.

albertshau replied:
In general I don't think it's great to assert exception messages in tests as it is very brittle and tends to just test how things happen to be implemented rather than testing an explicit contract. 

That is why I suggested catching the InvalidConfigPropertyException and asserting that the property name is correct. That is an actual contract that the plugin must follow.

For other scenarios it may make sense to do what you suggested.

mlozbin-cybervisiontech replied:
I was trying to say about failure messages in test logs/reports, not about asserting exception messages

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -0,0 +1,134 @@
+    );
+
+    Assertions.assertThat(config.resolveProject())
+      .isEqualTo(ServiceOptions.getDefaultProjectId());
```
Tests should use plain junit assertions when possible. Does this give you anything more than normal junit Assert.assertEquals()?

mlozbin-cybervisiontech replied:
> just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

mlozbin-cybervisiontech replied:
Actually assertEquals is almost the only case where JUnit assertion works in same way as AssertJ one. And even in this case there are some differences:
- JUnit assertions provoke developer to pass "expected" and "actual" params in wrong order and you'll have confusing failure message. This is fine when test failure is reproducible but often leads to loss of time when bug is reproducible only on CI or once per week (for example).
- JUnit's assertEquals does not have custom representation, custom comparator and soft assertions support.
- JUnit's assertEquals is less smart than one from AssertJ in some cases. For example: if you compare numbers with floating point assertJ will use comparison delta by default.

I also wanted to add why I prefer AssertJ in general:
- it provides much better failure messages. For example with JUnit:
    - you often use assertNotNull that will fail with AssertionError with no message (no details at all)
    - with code like Assert.assertTrue(one.getSchema().getFields().size() == two.getSchema().getFields().size()) you will not see wrong actual schema to analyze what happened. This is a problem when bug is not reproducible.
- it is single library that covers all possible cases where you previously needed Junit, TestNG or Hamcrest assertion libraries.
- it makes code much more readable and concise.
- it provides custom assertions functionality and other extendability options that allows to make testing frameworks. That approach reduces code duplication and makes all tests look similar without complex guidelines.
- it gives great auto-completion for IDE which is very handy with custom assertions.
- more detailed key difference overview: https://professional-practical-programmer.com/2016/06/26/assert-j/
- AsserJ features highlight: http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html

In summary. There are definitely some cases where AssertJ does better job than Junit assertions. And in this case there is a choice: to use 3 libraries depending on use case (Junit, hamcrest, AssertJ) or to use only 1 library in all cases (AssertJ). As for me, second approach is easier to describe in guides and to understand than trying to describe what library should be used in what case.

Please, consider this as my personal opinion and just a proposal. Thank you for your time.

albertshau replied:
I agree that junit doesn't have the most features, but I think other more powerful testing frameworks can sometimes allow you to get away with not having the best abstractions in the actual code. So I hesitate to add it where it is not necessary. As a general rule, I think it makes sense to start with only JUnit, since that is the standard across the other repositories, and only add others if there is a compelling reason to do it. In this particular case, it doesn't seem to give us anything that you couldn't do in junit.

mlozbin-cybervisiontech replied:
fixed

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -0,0 +1,134 @@
+      null
+    );
+
+    Assertions.assertThat(config.resolveServiceAccountFilePath())
```
Similar question here

mlozbin-cybervisiontech replied:
fixed

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -0,0 +1,134 @@
+package co.cask.cdc.plugins.sink;
```
missing copyright

mlozbin-cybervisiontech replied:
fixed

albertshau reviewed (commented):
just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

mlozbin-cybervisiontech commented:
> just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?

@albertshau This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

mlozbin-cybervisiontech committed:
CDAP-13558 Support CDC pipelines on Spark version 2.x
- implemented integration tests for CTSQLServer source plugin
- introduced OperationType enum to make CDC records more type safe

albertshau commented on src/test/java/co/cask/cdc/plugins/sink/CDCBigTableConfigTest.java:
```diff
@@ -36,7 +55,7 @@ public void testValidateReference() {
 
     Assertions.assertThatThrownBy(config::validate)
-      .isInstanceOf(InvalidConfigPropertyException.class);
+      .isInstanceOfSatisfying(InvalidConfigPropertyException.class, withProperty(Constants.Reference.REFERENCE_NAME));
```
It's simpler to just do:

```
try {
  config.validate()
} catch (InvalidConfigPropertyException e) {
  Assert.assertEquals(Constants.Reference.REFERENCE_NAME, e.getProperty());
}
```

this is just normal java that mirrors how the caller is expected to behave.

mlozbin-cybervisiontech replied:
fixed

mlozbin-cybervisiontech committed:
CDAP-13558 Support CDC pipelines on Spark version 2.x
- implemented integration tests for CTSQLServer source plugin
- introduced OperationType enum to make CDC records more type safe

mlozbin-cybervisiontech commented:
@albertshau, addressed all recent comments.

mlozbin-cybervisiontech commented:
> > just noticed that a lot of files are missing copyright. This is supposed to be in the rat check, have we not set it up for this project?
> 
> @albertshau This plugin was not configured for this project as well as for google-cloud. I may submit another PR with rat plugin and updated copyrights in all classes.

created https://github.com/data-integrations/change-data-capture/pull/9

albertshau reviewed (approved):
lgtm
